File: backend/app/models/project.py
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Index, String, Text, text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Project(Base):
    __tablename__ = "projects"
    __table_args__ = (Index("idx_projects_owner_user_id", "owner_user_id"),)

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False),
        primary_key=True,
        server_default=text("gen_random_uuid()"),
    )
    owner_user_id: Mapped[str] = mapped_column(
        UUID(as_uuid=False),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text(), nullable=True)
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, server_default=text("'active'")
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("now()")
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=text("now()")
    )

    owner = relationship("User", back_populates="projects")
    floors = relationship("Floor", back_populates="project")
    assets = relationship("Asset", back_populates="project")
    scene_drafts = relationship("SceneDraft", back_populates="project")

File: frontend/src/types/project.ts
import type { ISODateString, UUID } from './common';

export type ProjectStatus = 'active' | 'archived';

export interface Project {
  id: UUID;
  owner_user_id: UUID;
  name: string;
  description?: string | null;
  status: ProjectStatus;
  created_at: ISODateString;
  updated_at: ISODateString;
}

export interface CreateProjectRequest {
  name: string;
  description?: string;
}

export interface UpdateProjectRequest {
  name?: string;
  description?: string;
  status?: ProjectStatus;
}
